browser/lifetime/brave_application_lifetime_mac.mm

```diff
@@ -0,0 +1,33 @@
+/* Copyright (c) 2022 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/browser/lifetime/brave_application_lifetime_mac.h"
+
+#include "brave/browser/sparkle_buildflags.h"
+
+#if BUILDFLAG(ENABLE_SPARKLE)
+#import "brave/browser/mac/sparkle_glue.h"
+#endif
+
+namespace brave {
+
+bool AttemptRestartOnMac() {
+#if BUILDFLAG(ENABLE_SPARKLE)
+  if (!sparkle_glue::SparkleEnabled())
+    return false;
+
+  SparkleGlue* sparkle_glue = [SparkleGlue sharedSparkleGlue];
+  // Only relaunch via sparkle when update is available.
+  if ([sparkle_glue recentStatus] == kAutoupdateInstalled) {
+    [sparkle_glue relaunch];
```

Comment:
is there a reason to do this via sparkle? why can't we always restart via chromium's `AttemptRestart()` call?

It looks like this restart is pretty crazy for chromium (it kills the process in a non-chromium friendly way). I think the right thing here should be is to do this sparkle call as the last step `AttemptRestartInternal` (put this condition there).

Because we don't call `AttemptRestartInternal()` currently on MacOS updates (via Settings/Hamburger), we may lose sessions, break local state/prefs and all other crazy things.

For example, right now [this task](https://github.com/brave/brave-browser/issues/28379) is failing on MacOS most likely because `prefs::kWasRestarted` is not set during the restart triggered via a manual "Relaunch" click on `brave://settings/help`.

@simonhong we need to resurrect this PR and make it work as intended. I'll be happy to review it.

Reply:
Sparkle relaunch was introduced for fixing https://github.com/brave/brave-browser/issues/1787.
When asking relaunch to sparkle, not sure but I think sparkle does its things and sends normal quit signals as you investigated.
Yeah, doing it as a last step of `AttemptRestartInternal()` could be more better 👍🏼

Reply:
> Sparkle relaunch was introduced for fixing [brave/brave-browser#1787](https://github.com/brave/brave-browser/issues/1787). When asking relaunch to sparkle, not sure but I think sparkle does its things and sends normal quit signals as you investigated. Yeah, doing it as a last step of `AttemptRestartInternal()` could be more better 👍🏼

well, we might have fixed one issue (or maybe just hid it under the rug), but created another one.